File: testing/postgres-client-tests/java/PostgresTest.java
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.ResultSet;

public class PostgresTest {
    public String query;
    public Integer expectedUpdateCount; // for queries that don't return set of results
    public Object fieldAccessor; // can be String (i.e column name), Integer (i.e. field position) and `null` for no result set queries.
    public String[] expectedResults;

    // Parameterized constructor
    public PostgresTest(String query, Integer expectedUpdateCount, Object fieldAccessor, String[] expectedResults) {
        this.query = query;
        this.expectedUpdateCount = expectedUpdateCount;
        this.fieldAccessor = fieldAccessor;
        this.expectedResults = expectedResults;
    }
    // Getters and Setters (optional)
    public String getQuery() {
        return this.query;
    }

    public Integer getExpectedUpdateCount() {
        return this.expectedUpdateCount;
    }

    public Object getFieldAccessor() {
        return this.fieldAccessor;
    }

    public String[] getExpectedResults() {
        return this.expectedResults;
    }

    // test queries to be run against Doltgres
    private static final PostgresTest[] tests = {
        new PostgresTest("create table test (pk int, value int, d1 decimal(4,2), c1 char(10), primary key(pk))", 0, null, null),
        new PostgresTest("select pk from test", null, "pk", new String[]{}), // the table has no rows
        new PostgresTest("insert into test (pk, value, d1, c1) values (0,1,2.3,'hi'), (2,3,4.56,'hello')", 2, null, null),
        new PostgresTest("select * from test", null, "pk", new String[]{"0","2"}),
        new PostgresTest("select * from test", null, "value", new String[]{"1","3"}),
        // TODO: doltgres DECIMAL type result is returned as "2.3", should be "2.30"
        // new PostgresTest("select * from test", null, "d1", new String[]{"2.30","4.56"}),
        new PostgresTest("select * from test", null, "c1", new String[]{"hi        ","hello     "}),
        new PostgresTest("call dolt_add('-A')", 0, null, null),
        new PostgresTest("call dolt_commit('-m', 'my commit')", 0, null, null),
        new PostgresTest("select COUNT(*) FROM dolt_log", null, 1, new String[]{"2"}),
        new PostgresTest("call dolt_checkout('-b', 'mybranch')", 0, null, null),
        new PostgresTest("insert into test (pk, value, d1, c1) values (1,1,12.34,'bye')", 1, null, null),
        new PostgresTest("call dolt_commit('-a', '-m', 'my commit2')", 0, null, null),
        new PostgresTest("call dolt_checkout('main')", 0, null, null),
        new PostgresTest("call dolt_merge('mybranch')", 0, null, null),
        new PostgresTest("select COUNT(*) FROM dolt_log", null, "COUNT(*)", new String[]{"3"}), // returns res
    };

    public static void main(String[] args) {
        Connection conn = null;

        String user = args[0];
        String port = args[1];

        try {
            String url = "jdbc:postgresql://127.0.0.1:" + port + "/doltgres";
            String password = "";

            conn = DriverManager.getConnection(url, user, password);
            Statement st = conn.createStatement(ResultSet.TYPE_SCROLL_INSENSITIVE, ResultSet.CONCUR_READ_ONLY);

            for (int i = 0; i < tests.length; i++) {
                PostgresTest t = tests[i];
                String query    = t.getQuery();
                if ( st.execute(query) ) {
                    String[] expectedResults = t.getExpectedResults();
                    Object fieldAccessor = t.getFieldAccessor();
                    ResultSet rs = st.getResultSet();
                    int rowCount = getRowCount(rs);
                    // Compare the row count to the length of the array
                    if (rowCount != expectedResults.length) {
                        System.out.println("Row count in ResultSet: " + rowCount + " does not match length of the expected results: " + expectedResults.length);
                        System.exit(1);
                    }

                    Integer j = 0;
                    if (rs.next()) {
                        String expected = expectedResults[j];
                        String result = "";
                        if (fieldAccessor instanceof String) {
                            result = rs.getString((String)fieldAccessor);
                        } else if (fieldAccessor instanceof Integer) {
                            result = rs.getString((Integer)fieldAccessor);
                        } else {
                            System.out.println("Unsupported field accessor value: " + fieldAccessor);
                            System.exit(1);
                        }

                        if (!expected.equals(result) && !(query.contains("dolt_commit")) && !(query.contains("dolt_merge"))) {
                            System.out.println("Query: \n" + query);
                            System.out.println("Expected:\n'" + expected + "'");
                            System.out.println("Result:\n'" + result + "'");
                            System.exit(1);
                        }
                        j++;
                    }
                } else {
                    Integer expectedUpdateCount = t.getExpectedUpdateCount();
                    Integer result = st.getUpdateCount();
                    if ( !expectedUpdateCount.equals(result) ) {
                        System.out.println("Query: \n" + query);
                        System.out.println("Expected:\n" + expectedUpdateCount);
                        System.out.println("Rows Updated:\n" + result);
                        System.exit(1);
                    }
                }
            }
            System.exit(0);
        } catch (SQLException ex) {
            System.out.println("An error occurred.");
            ex.printStackTrace();
            System.exit(1);
        }
    }

    public static int getRowCount(ResultSet rs) throws SQLException {
        int rowCount = 0;
        if (rs != null) {
            rs.last(); // Move to the last row
            rowCount = rs.getRow(); // Get the row number (which is the row count)
            rs.beforeFirst(); // Move back to the beginning of the ResultSet
        }
        return rowCount;
    }
}

File: testing/postgres-client-tests/perl/postgres-test.pl
use strict;

use DBI;

my $QUERY_RESPONSE = [
    { "create table test (pk int, value int, d1 decimal(9, 3), f1 float, c1 char(10), t1 text, primary key(pk))" => '0E0' },
    { "insert into test (pk, value, d1, f1, c1, t1) values (0,0,0.0,0.0,'abc','a1')" => 1 },
    { "select * from test" => 1 },
    {"call dolt_add('-A');" => '0E0' },
    {"call dolt_commit('-m', 'my commit')" => '0E0' },
    {"call dolt_checkout('-b', 'mybranch')" => '0E0' },
    {"insert into test (pk, value, d1, f1, c1, t1) values (10,10, 123456.789, 420.42,'example','some text')" => 1 },
    {"call dolt_commit('-a', '-m', 'my commit2')" => '0E0' },
    {"call dolt_checkout('main')" => '0E0' },
    {"call dolt_merge('mybranch')" => '0E0' },
    {"select COUNT(*) FROM dolt.log" => 1 },
];

my $user = $ARGV[0];
my $port = $ARGV[1];
my $db   = "doltgres";

my $dsn = "DBI:Pg:database=$db;host=127.0.0.1;port=$port";
# Connect to the database
my $dbh = DBI->connect($dsn, $user, "password", { PrintError => 0, RaiseError => 1 });
die "failed to connect to database:DBI->errstr()" unless($dbh);

foreach my $query_response ( @{$QUERY_RESPONSE} ) {
    my @query_keys = keys %{$query_response};
    my $query      = $query_keys[0];
    my $exp_result = $query_response->{$query};

    my $result = $dbh->do($query);
    if ( $result != $exp_result ) {
	print "QUERY: $query\n";
	print "EXPECTED: $exp_result\n";
	print "RESULT: $result\n";
	exit 1
    }
}

# Define the SQL query
my $query = "SELECT * FROM test WHERE pk = 10";

# Prepare the query
my $sth = $dbh->prepare($query);

# Execute the query
$sth->execute();

my @cols = ("pk", "value", "d1", "f1", "c1", "t1");
my @expectedRowResults = ("10", "10", "123456.789", "420.42", "example   ", "some text");

# Fetch and process the results
while (my $row = $sth->fetchrow_hashref()) {
    if ($row) {
        # Process the row if it's defined
        # Access individual column values using hash dereferencing
        for my $i (0 .. $#cols) {
            my $val = $row->{$cols[$i]};
            # Comparing the strings using eq operator
            my $c = $val eq $expectedRowResults[$i];
            if ( $c == 0 ) {
                print "Expected: '$expectedRowResults[$i]'\n";
                print "Actual: '$val'\n";
                # Disconnect from the database
                $dbh->disconnect();
                exit 1;
            }
        }
    } else {
        print("no rows?");
        last; # Exit the loop when there are no more rows
    }
}

# Disconnect from the database
$dbh->disconnect();

exit 0;
